File: foundational/observability/framework/src/main/java/com/progressive/minds/chimera/core/observability/SparkMetricCollector.scala
package com.progressive.minds.chimera.core.observability

import org.apache.spark.sql.SparkSession
import org.apache.spark.scheduler.{SparkListener, SparkListenerJobEnd, SparkListenerJobStart, SparkListenerTaskEnd}
import scala.collection.mutable

class SparkMetricCollector(spark: SparkSession) extends SparkListener with MetricCollector {

  private val metrics: mutable.Map[String, Any] = mutable.Map.empty

  // Register this collector as a Spark listener
  spark.sparkContext.addSparkListener(this)

  override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
    metrics += s"job-${jobStart.jobId}-startTime" -> System.currentTimeMillis()
  }

  override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
    val jobEndTime = System.currentTimeMillis()
    metrics += s"job-${jobEnd.jobId}-endTime" -> jobEndTime
    val jobStartTime = metrics.getOrElse(s"job-${jobEnd.jobId}-startTime", jobEndTime).asInstanceOf[Long]
    metrics += s"job-${jobEnd.jobId}-duration" -> (jobEndTime - jobStartTime)
  }

  override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
    val peakMemory = taskEnd.taskMetrics.peakExecutionMemory
    metrics += s"task-${taskEnd.taskInfo.taskId}-peakMemory" -> peakMemory
  }

  override def collectMetric(name: String, value: Any): Unit = {
    metrics += name -> value
  }

  override def report(): Unit = {
    metrics.foreach { case (key, value) => println(s"$key: $value") }
  }

  // Expose metrics for testing
  def getMetrics: Map[String, Any] = metrics.toMap
}

File: core/observability/src/main/java/com/progressive/minds/chimera/SparkMetricCollector.java
package com.progressive.minds.chimera;

import org.apache.spark.scheduler.*;
import org.apache.spark.sql.SparkSession;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class SparkMetricCollector extends SparkListener implements MetricCollector {

    private final SparkSession sparkSession;
    private final Map<String, Object> metrics = Collections.synchronizedMap(new HashMap<>());

    public SparkMetricCollector(SparkSession sparkSession) {
        this.sparkSession = sparkSession;
        sparkSession.sparkContext().addSparkListener(this);
    }

    @Override
    public void onJobStart(SparkListenerJobStart jobStart) {
        metrics.put("job-" + jobStart.jobId() + "-startTime", System.currentTimeMillis());
    }

    @Override
    public void onJobEnd(SparkListenerJobEnd jobEnd) {
        long endTime = System.currentTimeMillis();
        metrics.put("job-" + jobEnd.jobId() + "-endTime", endTime);

        Long startTime = (Long) metrics.get("job-" + jobEnd.jobId() + "-startTime");
        if (startTime != null) {
            metrics.put("job-" + jobEnd.jobId() + "-duration", endTime - startTime);
        }
    }

    @Override
    public void onTaskEnd(SparkListenerTaskEnd taskEnd) {
        long peakMemory = taskEnd.taskMetrics().peakExecutionMemory();
        metrics.put("task-" + taskEnd.taskInfo().taskId() + "-peakMemory", peakMemory);
    }

    @Override
    public void collectMetric(String name, Object value) {
        metrics.put(name, value);
    }

    @Override
    public void report() {
        synchronized (metrics) {
            metrics.forEach((key, value) -> System.out.println(key + ": " + value));
        }
    }

    public Map<String, Object> getMetrics() {
        return new HashMap<>(metrics); // Return a copy to ensure immutability for external callers
    }
}
